### ISSUE TITLE
Can you explain this codecov report?

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#6888]
See https://codecov.io/gh/certbot/certbot/pull/6784.

Why is there a drop in coverage on the manual, nginx, and standalone plugins? Those plugins aren't at all affected by the PR.

Is there a way to see what codecov is comparing the coverage against? I spent a few minutes poking around and couldn't quickly figure it out.

**Comment by adferrand** [comment]
At this point, the status of this codecov report is correct in term of coverage, and in term of diff. I mean that there is no drop in the coverage, and the diff corresponds to what has been modified in this PR. The only strange result is about the list of changes. Several of them, untouched by the PR, are referenced in this list.

I suspect that the drop in coverage/diff was transient, due to missing reports from Windows or Linux tests that was not yet processed. After everything is aggregated, the coverage is normal. This specific issue is corrected by #6934

About the list of files, one can see that the real URL of https://codecov.io/gh/certbot/certbot/pull/6784 is https://codecov.io/gh/certbot/certbot/compare/491d6c8f4538c28f7c05d780db9a700982460242...3d811a784dba6f6a13fb773e35efc79a9aa0f867.

This URL gives use the two commits that are compared:
* `3d811a784dba6f6a13fb773e35efc79a9aa0f867` is the last commit of  `fix_centos6_ssl`, the merged branch by the PR (commit from 02/04/2019)
* `491d6c8f4538c28f7c05d780db9a700982460242` is the last ancestor on the master branch for the `fix_centos6_ssl` branch (commit from 27/03/2019)

Now, let's consider the last commit on master that was present when the codecov report has been triggered (so on 02/04/2019): it is `1daa3ca07634168945fc35ff44c413c6c94b795d` (the immediate commit before the one that integrated #6784.

And now so, let's compare `491d6c8f4538c28f7c05d780db9a700982460242` (the last ancestor for PR, so the based reference) and `1daa3ca07634168945fc35ff44c413c6c94b795d` (the latest commit in master at the time of the report generation):
https://github.com/certbot/certbot/compare/491d6c8f..1daa3ca0

The list of files given by this GitHub diff **is** the list of changes shown in the Codecov report changes https://codecov.io/gh/certbot/certbot/pull/6784/changes.

**Comment by adferrand** [comment]
So, to sum up.

1) The base commit on which codecov is constructing a comparison to build the report is the last ancestor on the target branch for the PR.
2) **BUT** the "changes" list displayed by Codecov for the PR is a comparison between the source branche and the actual state of the target branch, so its last commit at the time of the report.

So as a conclusion: to have a coherent and readable report, follow the always good practice to merge your source branch from master _just before_ the review. This will make the last ancestor of the target branch for a PR being also the last commit in the target branch.

**Comment by bmw** [comment]
> At this point, the status of this codecov report is correct in term of coverage, and in term of diff. I mean that there is no drop in the coverage, and the diff corresponds to what has been modified in this PR. 

I hadn't realized this report had changed out from underneath us. The report I was initially referring to that was available at the PR URL on codecov was https://codecov.io/gh/certbot/certbot/compare/491d6c8f4538c28f7c05d780db9a700982460242...f01a1a6fc7a3409022e92c03576e51b673c3fa69.

I think your explanation still applies here though. As you can see by the new link, the commits being compared are the latest commit in the PR and the shared ancestor of the PR and the target branch. The fact that the shared ancestor is old causes all sorts of other changes that have landed in `master` to creep into the report.

Do you know if there is a way we can change this? Can we tell codecov to use the latest commit it has seen in the target branch rather than the latest shared ancestor?

Alternatively we could make no changes and just remind everyone to update their PRs when they hit this problem. Yet another thing we could do is turn on a setting in GH requiring that PRs be up-to-date before merging, but I personally think that we have too many open PRs and too many people working on Certbot so making this change would be really painful.

**Comment by adferrand** [comment]
The documentation of codecov about this (https://docs.codecov.io/docs/comparing-commits and https://docs.codecov.io/docs/commit-status) does not indicate that there is a way to configure which commit will be used.

I think the best way is at least to make a strong advice about maintaining your PR up-to-date. Personally, I would even enforce it, since it ensures that the PR builds will correspond to the effective status of Certbot after the merge. What is painful about this is if you have conflicts with the base branch, but if it is the case the PR is already unmergeable, and a conflicts must be solved using a merge from master.

**Comment by bmw** [comment]
I agree there's value in having every PR be up-to-date before merging for the sake of testing, but I personally don't think it's worth it.

I disagree that the main pain point is merge conflicts. If we enforced that every PR Is up-to-date before merging, as soon as one PR lands, every other PR must be updated before it can be merged. I think this requires the author and the review of the PR to work together in quick succession or risk the chance that another PR will be merged requiring the author to update the PR again.

For me, given the choice between that and these codecov issues, I'd personally take the codecov issues.

It probably doesn't make sense for us to look into making this transition now, but I wonder how [coveralls](https://coveralls.io/) deals with this and some of the other problems we've been having with codecov. It's been a while since we used it, but I don't remember problems like this.
